// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: examples/echo/v1/service.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on SpeakRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SpeakRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SpeakRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in SpeakRequestMultiError, or
// nil if none found.
func (m *SpeakRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *SpeakRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Input

	if len(errors) > 0 {
		return SpeakRequestMultiError(errors)
	}

	return nil
}

// SpeakRequestMultiError is an error wrapping multiple validation errors
// returned by SpeakRequest.ValidateAll() if the designated constraints aren't met.
type SpeakRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SpeakRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SpeakRequestMultiError) AllErrors() []error { return m }

// SpeakRequestValidationError is the validation error returned by
// SpeakRequest.Validate if the designated constraints aren't met.
type SpeakRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SpeakRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SpeakRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SpeakRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SpeakRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SpeakRequestValidationError) ErrorName() string { return "SpeakRequestValidationError" }

// Error satisfies the builtin error interface
func (e SpeakRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSpeakRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SpeakRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SpeakRequestValidationError{}

// Validate checks the field values on SpeakResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SpeakResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SpeakResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in SpeakResponseMultiError, or
// nil if none found.
func (m *SpeakResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *SpeakResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Output

	if len(errors) > 0 {
		return SpeakResponseMultiError(errors)
	}

	return nil
}

// SpeakResponseMultiError is an error wrapping multiple validation errors
// returned by SpeakResponse.ValidateAll() if the designated constraints
// aren't met.
type SpeakResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SpeakResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SpeakResponseMultiError) AllErrors() []error { return m }

// SpeakResponseValidationError is the validation error returned by
// SpeakResponse.Validate if the designated constraints aren't met.
type SpeakResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SpeakResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SpeakResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SpeakResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SpeakResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SpeakResponseValidationError) ErrorName() string { return "SpeakResponseValidationError" }

// Error satisfies the builtin error interface
func (e SpeakResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSpeakResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SpeakResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SpeakResponseValidationError{}
